analyzer/src/main/kotlin/managers/Pub.kt

```diff
@@ -0,0 +1,538 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.analyzer.managers
+
+import ch.frankel.slf4k.*
+
+import com.fasterxml.jackson.databind.JsonNode
+
+import com.here.ort.analyzer.AbstractPackageManagerFactory
+import com.here.ort.analyzer.PackageManager
+import com.here.ort.downloader.VersionControlSystem
+import com.here.ort.model.Identifier
+import com.here.ort.model.Package
+import com.here.ort.model.PackageReference
+import com.here.ort.model.Project
+import com.here.ort.model.ProjectAnalyzerResult
+import com.here.ort.model.RemoteArtifact
+import com.here.ort.model.Scope
+import com.here.ort.model.yamlMapper
+import com.here.ort.model.VcsInfo
+import com.here.ort.model.config.AnalyzerConfiguration
+import com.here.ort.model.config.RepositoryConfiguration
+import com.here.ort.model.OrtIssue
+
+import com.here.ort.utils.CommandLineTool
+import com.here.ort.utils.ProcessCapture
+import com.here.ort.utils.log
+import com.here.ort.utils.showStackTrace
+import com.here.ort.utils.textValueOrEmpty
+import com.here.ort.utils.collectMessagesAsString
+import com.here.ort.utils.getUserHomeDirectory
+import com.vdurmont.semver4j.Requirement
+
+import java.io.File
+import java.io.IOException
+import java.util.SortedSet
+
+const val PUB_LOCK_FILE = "pubspec.lock"
+
+/**
+ * The [Pub](https://https://pub.dev/) package manager for Dart / Flutter.
+ */
+class Pub (
+    name: String,
+    analyzerRoot: File,
+    analyzerConfig: AnalyzerConfiguration,
+    repoConfig: RepositoryConfiguration
+) : PackageManager(name, analyzerRoot, analyzerConfig, repoConfig), CommandLineTool {
+    class Factory : AbstractPackageManagerFactory<Pub>("Pub") {
+        override val globsForDefinitionFiles = listOf("pubspec.yaml")
+
+        override fun create(
+            analyzerRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) =
+            Pub(managerName, analyzerRoot, analyzerConfig, repoConfig)
+    }
+
+    override fun getVersionRequirement(): Requirement = Requirement.buildIvy("[2.2,)")
+
+    override fun beforeResolution(definitionFiles: List<File>) {
+
+    }
+
+    /**
+     * A workspace reader that is backed by the local Gradle artifact cache.
+     */
+    private class PubCacheReader {
+        // TODO: Check if we can support a second .pub-cache
+        // If there is a .pub-cache directory in Flutter root,
+        // flutter packages get will use that one instead of the .pub-cache directory in user home
+        // Currently we do not support usage of two different pub-cache directories.
+        // Therefore you need to delete the .pub-cache directory in your flutter root folder
+        // to force flutter to use the default one
```

Comment:
Please use the max allowed line length of 120 for this comment, to make it take less lines.